### ISSUE TITLE
Python global dataflow analysis do not track @classmethod correctly.

### ISSUE THREAD

**Issue by hksdpc255** [github/codeql#16953]
## Python
```python
class A:
    def __init__(self, num) -> None:
        print(num)

class B(A):
    def __init__(self) -> None:
        super().__init__(1)

    @classmethod
    def create(cls, num2=2):
        return cls(num2)


if __name__ == '__main__':
    B.create(3)
```
## CodeQL
```codeql
import python
import semmle.python.ApiGraphs
import semmle.python.dataflow.new.DataFlow
import semmle.python.dataflow.new.TaintTracking

module MyConf implements DataFlow::ConfigSig {
    predicate isSource(DataFlow::Node source) {
        source.asExpr() instanceof IntegerLiteral
    }
    predicate isSink(DataFlow::Node sink) {
        sink = API::builtin("print").getACall().getArg(0)
    }
}

module MyFlow = DataFlow::Global<MyConf>;

from DataFlow::Node source, DataFlow::Node sink
where MyFlow::flow(source, sink)
select source, sink
```
## Query result
|source|sink|
|-|-|
|1|num|
## Expected result
|source|sink|
|-|-|
|1|num|
|2|num|
|3|num|

**Comment by tausbn** [comment]
Thank you for your report. I must admit, I'm slightly confused as to what you're trying to do here.

First of all: your code doesn't run. The call to `B.create(3)` results in the following exception:
```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 6, in create
TypeError: B.__init__() takes 1 positional argument but 2 were given
```

But let's assume you meant to write 
```python
class B(A):
    def __init__(self, num):
        super().__init__(1)
```
In this case, we can run the code, and we get a printout of... `1` (because that's what you're ultimately passing into `A.__init__` where it gets printed).

So I'm really not sure why you expect flow from `2` and `3` in this case. For both of these, the flow stops at the call `cls(num2)` (in the broken version) or the `num` parameter of `B.__init__` (in the "fix" I propose above).

If we also change the `super().__init__(1)` call to be `super().__init__(num)`, then we get the expected flow from both `2` and `3` (but not `1`, of course, as we just got rid of that source).

In conclusion, I think everything is working as intended?

**Comment by hksdpc255** [comment]
Sorry for my mistake. `super().__init__(num)` is exactly what I meant, and CodeQL does capture the expected flow from both `2` and `3`.

I'm writing a CodeQL query to look up a function argument's possible values.

I'm currently writing a CodeQL query to identify a function argument's possible values. Specifically, I am trying to determine if the `topic=None` located in this file:
https://github.com/openstack/neutron/blob/a8bdcc768a92ac7ddd2eb911f20e2804e9922088/neutron/service.py#L416
will be passed into `oslo_messaging.get_rpc_server(..., target, ...)`. However, I've been unsuccessful in this endeavor using CodeQL's built-in global dataflow analysis.

The expected flow will be:
1. `topic=None` in https://github.com/openstack/neutron/blob/a8bdcc768a92ac7ddd2eb911f20e2804e9922088/neutron/service.py#L416
2. `topic` in https://github.com/openstack/neutron/blob/a8bdcc768a92ac7ddd2eb911f20e2804e9922088/neutron/service.py#L445
3. `topic` in https://github.com/openstack/neutron/blob/a8bdcc768a92ac7ddd2eb911f20e2804e9922088/neutron/service.py#L372C38-L372C43
4. `topic` in https://github.com/openstack/neutron/blob/a8bdcc768a92ac7ddd2eb911f20e2804e9922088/neutron/service.py#L386
5. `topic` in https://github.com/openstack/neutron-lib/blob/47e1736c10ffd6f1850eba9970acdc00373f9de7/neutron_lib/rpc.py#L291
6. `self.topic` in https://github.com/openstack/neutron-lib/blob/47e1736c10ffd6f1850eba9970acdc00373f9de7/neutron_lib/rpc.py#L294
7. `self.topic` in https://github.com/openstack/neutron-lib/blob/47e1736c10ffd6f1850eba9970acdc00373f9de7/neutron_lib/rpc.py#L310
8. `topic` in https://github.com/openstack/neutron-lib/blob/47e1736c10ffd6f1850eba9970acdc00373f9de7/neutron_lib/rpc.py#L337
9. `topic` in https://github.com/openstack/neutron-lib/blob/47e1736c10ffd6f1850eba9970acdc00373f9de7/neutron_lib/rpc.py#L339
10. `target` in https://github.com/openstack/neutron-lib/blob/47e1736c10ffd6f1850eba9970acdc00373f9de7/neutron_lib/rpc.py#L338 \( I did an additional flow step from `topic` to `Target(..., topic=topic, ...)`. \)
11. `target` in https://github.com/openstack/neutron-lib/blob/47e1736c10ffd6f1850eba9970acdc00373f9de7/neutron_lib/rpc.py#L340
12. `target` in https://github.com/openstack/neutron-lib/blob/47e1736c10ffd6f1850eba9970acdc00373f9de7/neutron_lib/rpc.py#L213
13. `target` in https://github.com/openstack/neutron-lib/blob/47e1736c10ffd6f1850eba9970acdc00373f9de7/neutron_lib/rpc.py#L224
14. Finally `oslo_messaging.get_rpc_server(..., target, ...)` in https://github.com/openstack/neutron-lib/blob/47e1736c10ffd6f1850eba9970acdc00373f9de7/neutron_lib/rpc.py#L224

This example was so complex that I wrote the wrong use case at the beginning while trying to find where it breaks, which led me to mistakenly believe that CodeQL cannot handle @classmethod correctly.

Could you help me investigate where the problem might be?

**Comment by tausbn** [comment]
> Could you help me investigate where the problem might be?

Alas, I do not have the time to help you out with such a broad issue.

I would suggest looking at where the flow stops when going forward from the source (and backwards from the sink).

For this, the documentation on [how to debug data-flow queries](https://codeql.github.com/docs/writing-codeql-queries/debugging-data-flow-queries-using-partial-flow/) may be of interest.
